src/peft/tuners/tuners_utils.py

```diff
@@ -665,3 +681,32 @@ def check_adapters_to_merge(module: BaseTunerLayer, adapter_names: Optional[list
             warnings.warn("All adapters are already merged, nothing to do.")
 
     return adapter_names
+
+
+def clone_module(module: nn.Module, share_weights=False):
+    clone = copy.deepcopy(module)
+
+    def _share_weights(src: nn.Module, dst: nn.Module):
+        for name, param in src.named_parameters(recurse=False):
+            dst.register_parameter(name, param)
+
+    if share_weights:
+        for name, submodule in module.named_modules():
+            _share_weights(submodule, clone.get_submodule(name))
+
+    return clone
+
+
+def replicate_layers(model: nn.Module, layer_map: List[Tuple[int, int]]):
+    new_layers = []
+    for start, end in layer_map:
+        for i in range(start, end):
+            current_idx = len(new_layers)
+            new_layers.append(clone_module(model.base_model.layers[i], share_weights=True))
+            # This is a hack needed to work around the layer_idx introduced in HF transformers.
+            for submodule in new_layers[-1].modules():
+                if hasattr(submodule, 'layer_idx'):
```

Comment:
What happens if we don't find this attribute on any submodule? Should we raise an error or is it okay to just ignore this?

Reply:
I think it is ok to ignore it. I have opened an issue on avoiding this layer_idx attribute so that layers are not tied to their position in the transformers repo. Have not had a chance to send a PR. Some transformers don't need the layer idx. In fact, the layer idx is primarily used for inference optimization with the key value cache. It does not impact training.